# Price data quality dashboard

## Imports and set up django environment

In [35]:
from datetime import timedelta
import math
import os
import pandas as pd
import django
from django.db import connection
from django.utils import timezone
import logging
import numpy as np
from IPython.core.display import display, HTML
os.chdir('..')

# Allows async calls to django ORM in Jupyter. Required.
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'rest.settings')
os.environ["DJANGO_ALLOW_ASYNC_UNSAFE"] = "true"
django.setup()

from pricedata import models





## Time ORM vs  Raw SQL
Retrieve data into dsataframe. We need:
* Symbol Name;
* Instrument Type Name;
* DataSource Name; and
* Time.

Where Retrieve Price Data is True and period is '1S'

## Django ORM QuerySet then convert to dataframe

In [32]:
%%timeit
price_data =  models.Candle.objects.filter(datasource_symbol__retrieve_price_data=True, period='1S')


df = pd.DataFrame(list(price_data.values('datasource_symbol__symbol__name',
                                                 'datasource_symbol__symbol__instrument_type',
                                                 'datasource_symbol__datasource__name', 'time')))

print(len(df.index))

2147012
2147012
2147012
2147012
2147012
2147012
2147012
2147012
25.4 s ± 1.45 s per loop (mean ± std. dev. of 7 runs, 1 loop each)


## Raw sql straight into dataframe


In [61]:
%%timeit
sql = """SELECT	s.name AS symbol,
                s.instrument_type AS instrument_type,
                ds.name AS datasource,
                cdl.time AS time
        FROM public.pricedata_datasourcesymbol dss
            INNER JOIN pricedata_datasource ds ON dss.datasource_id = ds.id
            INNER JOIN pricedata_symbol s ON dss.symbol_id = s.id
            INNER JOIN pricedata_candle cdl ON cdl.datasource_symbol_id = dss.id  
        WHERE dss.retrieve_price_data = true
            AND cdl.period = %(period)s"""
    
df = pd.read_sql_query(sql=sql, con=connection, params={'period': '1S'})

print(len(df.index))

2240566


## % difference between results

In [55]:
100 - 22.2 / 25.4 * 100

12.5984251968504

## Results
### Django
Time: 25.4 s ± 1.45 s per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rowcount: 2147012
    
### Raw
Time: 22.2 s ± 221 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
Rowcount:214012

#### We have a winner. Raw SQL is 12.5% faster than Django
    


## Get the data using raw SQL, this time without %%timeit

In [2]:
sql = """SELECT	s.name AS symbol,
                s.instrument_type AS instrument_type,
                ds.name AS datasource,
                cdl.time AS time
        FROM public.pricedata_datasourcesymbol dss
            INNER JOIN pricedata_datasource ds ON dss.datasource_id = ds.id
            INNER JOIN pricedata_symbol s ON dss.symbol_id = s.id
            INNER JOIN pricedata_candle cdl ON cdl.datasource_symbol_id = dss.id  
        WHERE dss.retrieve_price_data = true
            AND cdl.period = %(period)s"""
    
df = pd.read_sql_query(sql=sql, con=connection, params={'period': '1S'})

df

symbol instrument_type datasource                      time
0        CADCHF           FOREX        MT5 2021-07-16 00:00:00+00:00
1        CADCHF           FOREX        MT5 2021-07-16 00:05:01+00:00
2        CADCHF           FOREX        MT5 2021-07-16 00:05:27+00:00
3        CADCHF           FOREX        MT5 2021-07-16 00:05:43+00:00
4        CADCHF           FOREX        MT5 2021-07-16 00:05:51+00:00
...         ...             ...        ...                       ...
4285989  USDCNH           FOREX        MT5 2021-07-20 10:25:13+00:00
4285990  USDCNH           FOREX        MT5 2021-07-20 10:25:14+00:00
4285991  USDCNH           FOREX        MT5 2021-07-20 10:25:15+00:00
4285992  USDCNH           FOREX        MT5 2021-07-20 10:25:16+00:00
4285993  USDCNH           FOREX        MT5 2021-07-20 10:25:17+00:00

[4285994 rows x 4 columns]

## Add another datasource for testing

In [136]:
nds = df.copy()
nds['datasource'] = 'NDS'
df = df.append(nds)

print(f"NDS has {len(df[df['datasource'] == 'NDS'].index)} rows. MT5 has {len(df[df['datasource'] == 'MT5'].index)} rows.")

df

NDS has 2473058 rows. MT5 has 2473058 rows.


symbol instrument_type datasource                      time
0        CADCHF           FOREX        MT5 2021-07-16 00:00:00+00:00
1        CADCHF           FOREX        MT5 2021-07-16 00:05:01+00:00
2        CADCHF           FOREX        MT5 2021-07-16 00:05:27+00:00
3        CADCHF           FOREX        MT5 2021-07-16 00:05:43+00:00
4        CADCHF           FOREX        MT5 2021-07-16 00:05:51+00:00
...         ...             ...        ...                       ...
2473053  USDCNH           FOREX        NDS 2021-07-19 12:25:07+00:00
2473054  USDCNH           FOREX        NDS 2021-07-19 12:25:08+00:00
2473055  USDCNH           FOREX        NDS 2021-07-19 12:25:10+00:00
2473056  USDCNH           FOREX        NDS 2021-07-19 12:25:11+00:00
2473057  USDCNH           FOREX        NDS 2021-07-19 12:25:12+00:00

[4946116 rows x 4 columns]

## Table showing num prices by symbol, by DS for each aggregarion period

In [82]:
# First group by to get min and max times
grouped = df.groupby(['symbol', 'instrument_type', 'datasource']).agg(first=('time', 'min'), last=('time', 'max'), count=('time', 'count'))

# Now get min, max and mean for each aggregation period for each datasource.
# 'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'
aggs = {'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'}

for key in aggs:
    # Get counts for aggregation period, then group by symbol, instrument type and datasource to get min, max and avg counts for aggregation period
    agg_period_ungrouped = df.groupby(['symbol', 'instrument_type', 'datasource', pd.Grouper(key='time', freq=aggs[key])]).agg(count=('time', 'count'))
    agg_period_grouped = agg_period_ungrouped.groupby(['symbol', 'instrument_type', 'datasource']).agg(min=('count', 'min'), max=('count', 'max'), avg=('count', 'median'))

    # Rename columns to include aggregation period key, then merge into original dataframe
    agg_period_grouped = agg_period_grouped.rename(columns={'min': f'{key}_min', 'max': f'{key}_max', 'avg': f'{key}_avg'})
    grouped = grouped.join(agg_period_grouped, on=['symbol', 'instrument_type', 'datasource'])
    
# Now unstack, so we show columns for each datasource
grouped = grouped.unstack()

# Create the same aggregations but across all datasources
agg_cols = set([x[0] for x in grouped.columns])
ds_cols = set([x[1] for x in grouped.columns])

for agg_col in agg_cols:
    if '_min' in agg_col or agg_col == 'first':
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].min(axis=1)
    elif '_max' in agg_col or agg_col == 'last':
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].max(axis=1)
    elif '_avg' in agg_col:
        grouped[(agg_col, 'all')] = grouped[[(agg_col, ds) for ds in ds_cols]].mean(axis=1)
        

# Flatten the columns and reset the index
grouped.columns = ['_'.join(col).strip() for col in grouped.columns]
grouped = grouped.reset_index()
                          
grouped


symbol instrument_type                 first_MT5  \
0   AUDCAD           FOREX 2021-07-16 00:00:00+00:00   
1   AUDCHF           FOREX 2021-07-16 00:00:00+00:00   
2   AUDJPY           FOREX 2021-07-16 00:00:00+00:00   
3   AUDNZD           FOREX 2021-07-16 00:00:00+00:00   
4   AUDUSD           FOREX 2021-07-16 00:00:00+00:00   
5   CADCHF           FOREX 2021-07-16 00:00:00+00:00   
6   CADJPY           FOREX 2021-07-16 00:00:00+00:00   
7   CHFJPY           FOREX 2021-07-16 00:00:00+00:00   
8   CHFSGD           FOREX 2021-07-16 00:00:00+00:00   
9   EURAUD           FOREX 2021-07-16 00:00:00+00:00   
10  EURCAD           FOREX 2021-07-16 00:00:00+00:00   
11  EURCHF           FOREX 2021-07-16 00:00:00+00:00   
12  EURDKK           FOREX 2021-07-16 00:00:00+00:00   
13  EURGBP           FOREX 2021-07-16 00:00:00+00:00   
14  EURHKD           FOREX 2021-07-16 00:00:00+00:00   
15  EURHUF           FOREX 2021-07-16 00:00:00+00:00   
16  EURJPY           FOREX 2021-07-16 00:00:00+00:00   
17  EURNOK           FOREX 2021-07-16 00:00:00+00:00   
18  EURNZD           FOREX 2021-07-16 00:00:00+00:00   
19  EURPLN           FOREX 2021-07-16 00:00:00+00:00   
20  EURRUB           FOREX 2021-07-16 10:00:00+00:00   
21  EURSEK           FOREX 2021-07-16 00:00:00+00:00   
22  EURSGD           FOREX 2021-07-16 00:00:00+00:00   
23  EURTRY           FOREX 2021-07-16 00:00:00+00:00   
24  EURUSD           FOREX 2021-07-16 00:00:00+00:00   
25  EURZAR           FOREX 2021-07-16 00:00:00+00:00   
26  GBPAUD           FOREX 2021-07-16 00:00:00+00:00   
27  GBPCAD           FOREX 2021-07-16 00:00:00+00:00   
28  GBPCHF           FOREX 2021-07-16 00:00:00+00:00   
29  GBPDKK           FOREX 2021-07-16 00:00:00+00:00   
30  GBPJPY           FOREX 2021-07-16 00:00:00+00:00   
31  GBPNOK           FOREX 2021-07-16 00:00:00+00:00   
32  GBPNZD           FOREX 2021-07-16 00:00:00+00:00   
33  GBPSEK           FOREX 2021-07-16 00:00:00+00:00   
34  GBPSGD           FOREX 2021-07-16 00:00:00+00:00   
35  GBPUSD           FOREX 2021-07-16 00:00:00+00:00   
36  NZDCAD           FOREX 2021-07-16 00:00:00+00:00   
37  NZDCHF           FOREX 2021-07-16 00:00:00+00:00   
38  NZDJPY           FOREX 2021-07-16 00:00:00+00:00   
39  NZDSGD           FOREX 2021-07-16 00:00:00+00:00   
40  NZDUSD           FOREX 2021-07-16 00:00:00+00:00   
41  SGDJPY           FOREX 2021-07-16 00:00:00+00:00   
42  USDCAD           FOREX 2021-07-16 00:00:00+00:00   
43  USDCHF           FOREX 2021-07-16 00:00:00+00:00   
44  USDCNH           FOREX 2021-07-16 00:00:00+00:00   
45  USDDKK           FOREX 2021-07-16 00:00:00+00:00   
46  USDHKD           FOREX 2021-07-16 00:00:00+00:00   
47  USDHUF           FOREX 2021-07-16 00:00:00+00:00   
48  USDJPY           FOREX 2021-07-16 00:00:00+00:00   
49  USDMXN           FOREX 2021-07-16 00:00:00+00:00   
50  USDNOK           FOREX 2021-07-16 00:00:00+00:00   
51  USDPLN           FOREX 2021-07-16 00:00:00+00:00   
52  USDRUB           FOREX 2021-07-16 10:00:00+00:00   
53  USDSEK           FOREX 2021-07-16 00:00:00+00:00   
54  USDSGD           FOREX 2021-07-16 00:00:00+00:00   
55  USDTRY           FOREX 2021-07-16 00:00:00+00:00   
56  USDZAR           FOREX 2021-07-16 00:00:00+00:00   

                    last_MT5  count_MT5  minutes_min_MT5  minutes_max_MT5  \
0  2021-07-19 21:24:47+00:00      75431                1               60   
1  2021-07-19 21:24:51+00:00      71641                1               60   
2  2021-07-19 21:24:52+00:00      76895                1               60   
3  2021-07-19 21:24:47+00:00      72478                1               60   
4  2021-07-19 21:24:47+00:00      70547                1               60   
5  2021-07-19 21:24:23+00:00      54384                1               59   
6  2021-07-19 21:24:26+00:00      64055                1               59   
7  2021-07-19 21:24:23+00:00      58093                1               59   
8  2021-07-19 21:24:58+00:00      41969                1      

## We will keep the full data table intact, and create a new one for display purposes. This will:
* Remove the DS agg columns; 
* Move the agg functions into the cells; 
* Rename the columns; and
* Reorder columns.

In [83]:
display = grouped

# Merge the agg functions for _all into cells for each aggregation period
for agg in ['minutes', 'hours', 'days', 'weeks', 'months']:
    # Get the min, max and avg columns and consolidate into a new column. Drop the existing ones.
    min_col = f'{agg}_min_all'
    max_col = f'{agg}_max_all'
    avg_col = f'{agg}_avg_all'
    display[agg] = 'min:' + display[min_col].astype(str) + ' max:' + display[max_col].astype(str) + ' avg:' + display[avg_col].astype(str)

# Rename the columns that we will be keeping
display = display.rename(
    columns={'symbol': 'Symbol', 'instrument_type': 'Instrument Type', 'first_all': 'First Price',
             'last_all': 'Last Price', 'minutes': 'Minutes', 'hours': 'Hours', 'days': 'Days', 'weeks': 'Weeks',
             'months': 'Months'})

# Reorder the columns, excluding any that we dont want to keep
display = display[
    ['Symbol', 'Instrument Type', 'First Price', 'Last Price', 'Minutes', 'Hours', 'Days', 'Weeks', 'Months']]

display

Symbol Instrument Type               First Price  \
0   AUDCAD           FOREX 2021-07-16 00:00:00+00:00   
1   AUDCHF           FOREX 2021-07-16 00:00:00+00:00   
2   AUDJPY           FOREX 2021-07-16 00:00:00+00:00   
3   AUDNZD           FOREX 2021-07-16 00:00:00+00:00   
4   AUDUSD           FOREX 2021-07-16 00:00:00+00:00   
5   CADCHF           FOREX 2021-07-16 00:00:00+00:00   
6   CADJPY           FOREX 2021-07-16 00:00:00+00:00   
7   CHFJPY           FOREX 2021-07-16 00:00:00+00:00   
8   CHFSGD           FOREX 2021-07-16 00:00:00+00:00   
9   EURAUD           FOREX 2021-07-16 00:00:00+00:00   
10  EURCAD           FOREX 2021-07-16 00:00:00+00:00   
11  EURCHF           FOREX 2021-07-16 00:00:00+00:00   
12  EURDKK           FOREX 2021-07-16 00:00:00+00:00   
13  EURGBP           FOREX 2021-07-16 00:00:00+00:00   
14  EURHKD           FOREX 2021-07-16 00:00:00+00:00   
15  EURHUF           FOREX 2021-07-16 00:00:00+00:00   
16  EURJPY           FOREX 2021-07-16 00:00:00+00:00   
17  EURNOK           FOREX 2021-07-16 00:00:00+00:00   
18  EURNZD           FOREX 2021-07-16 00:00:00+00:00   
19  EURPLN           FOREX 2021-07-16 00:00:00+00:00   
20  EURRUB           FOREX 2021-07-16 10:00:00+00:00   
21  EURSEK           FOREX 2021-07-16 00:00:00+00:00   
22  EURSGD           FOREX 2021-07-16 00:00:00+00:00   
23  EURTRY           FOREX 2021-07-16 00:00:00+00:00   
24  EURUSD           FOREX 2021-07-16 00:00:00+00:00   
25  EURZAR           FOREX 2021-07-16 00:00:00+00:00   
26  GBPAUD           FOREX 2021-07-16 00:00:00+00:00   
27  GBPCAD           FOREX 2021-07-16 00:00:00+00:00   
28  GBPCHF           FOREX 2021-07-16 00:00:00+00:00   
29  GBPDKK           FOREX 2021-07-16 00:00:00+00:00   
30  GBPJPY           FOREX 2021-07-16 00:00:00+00:00   
31  GBPNOK           FOREX 2021-07-16 00:00:00+00:00   
32  GBPNZD           FOREX 2021-07-16 00:00:00+00:00   
33  GBPSEK           FOREX 2021-07-16 00:00:00+00:00   
34  GBPSGD           FOREX 2021-07-16 00:00:00+00:00   
35  GBPUSD           FOREX 2021-07-16 00:00:00+00:00   
36  NZDCAD           FOREX 2021-07-16 00:00:00+00:00   
37  NZDCHF           FOREX 2021-07-16 00:00:00+00:00   
38  NZDJPY           FOREX 2021-07-16 00:00:00+00:00   
39  NZDSGD           FOREX 2021-07-16 00:00:00+00:00   
40  NZDUSD           FOREX 2021-07-16 00:00:00+00:00   
41  SGDJPY           FOREX 2021-07-16 00:00:00+00:00   
42  USDCAD           FOREX 2021-07-16 00:00:00+00:00   
43  USDCHF           FOREX 2021-07-16 00:00:00+00:00   
44  USDCNH           FOREX 2021-07-16 00:00:00+00:00   
45  USDDKK           FOREX 2021-07-16 00:00:00+00:00   
46  USDHKD           FOREX 2021-07-16 00:00:00+00:00   
47  USDHUF           FOREX 2021-07-16 00:00:00+00:00   
48  USDJPY           FOREX 2021-07-16 00:00:00+00:00   
49  USDMXN           FOREX 2021-07-16 00:00:00+00:00   
50  USDNOK           FOREX 2021-07-16 00:00:00+00:00   
51  USDPLN           FOREX 2021-07-16 00:00:00+00:00   
52  USDRUB           FOREX 2021-07-16 10:00:00+00:00   
53  USDSEK           FOREX 2021-07-16 00:00:00+00:00   
54  USDSGD           FOREX 2021-07-16 00:00:00+00:00   
55  USDTRY           FOREX 2021-07-16 00:00:00+00:00   
56  USDZAR           FOREX 2021-07-16 00:00:00+00:00   

                  Last Price                 Minutes  \
0  2021-07-19 21:24:47+00:00   min:1 max:60 avg:27.0   
1  2021-07-19 21:24:51+00:00   min:1 max:60 avg:26.0   
2  2021-07-19 21:24:52+00:00   min:1 max:60 avg:27.0   
3  2021-07-19 21:24:47+00:00   min:1 max:60 avg:25.0   
4  2021-07-19 21:24:47+00:00   min:1 max:60 avg:24.0   
5  2021-07-19 21:24:23+00:00   min:1 max:59 avg:18.0   
6  2021-07-19 21:24:26+00:00   min:1 max:59 avg:22.0   
7  2021-07-19 21:24:23+00:00   min:1 max:59 avg:20.0   
8  2021-07-19 21:24:58+00:00   min:1 max:56 avg:14.0   
9  2021-07-19 21:24:55+00:00   min:1 max:60 avg:36.0   
10 2021-07-19 21:24:26+00:00   min:1 max:60 avg:32.0   
11 2021-07-19 21:24:19+00:00   min:1 max:55 avg:16.0   
12 2021-07-19 21:25:15+00:0

### This is slow. Timeit shows: 21.9 s ± 60.9 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

## Tieing it all together
## We will need a bulk upload method and will run this as a task

In [50]:
def bulk_insert_or_update(data: pd.DataFrame, table: str, unique_fields=None, batch_size=None):
    """
    Bulk insert or update (upsert) of price data. If unique fields already exists, then update else insert

    :param data: The pandas dataframe to insert / update to db. The columns in the dataframe must match the table
        columns.
    :param table: The name of the table to update
    :param unique_fields: Fields that will raise the unique key constraint on insert. If none are provided, then we 
        will just do a straight insert rather than upsert.
    :param batch_size: Maximum number of rows to update in one go. If None, then no batching
    :return:
    """

    # Logger
    log = logging.getLogger(__name__)
    
    # Do we have any data
    if data is not None and len(data.index) > 0:
        # Get create fields from dataframe and the update fields as the create fields - unique fields 
        create_fields = data.columns
        update_fields = set(create_fields) - set(unique_fields) if unique_fields is not None else None
        
        # Get batches if we are batching. If not create a single batch with all the data
        batches = [data, ] if not batch_size else np.array_split(data, math.ceil(len(data.index) / batch_size))
        log.info(f'Bulk update to {table}. Rowcount: {len(data.index)}. Update split into {len(batches)} batches of maximum {batch_size} updates.')  # TODO Debug
        
        for i in range(0 , len(batches)):
            batch = batches[i]
            log.info(f'Bulk update to {table}. Batch {i+1} of {len(batches)}.')  # TODO Debug
            
            # Get the values from the data
            values = [tuple(x) for x in batch.to_numpy()]

            # Create cursor
            cursor = connection.cursor()

            # Mogrify values to bind into sql.
            placeholders = ','.join(['%s' for _ in create_fields])
            mogvals = [cursor.mogrify(f"({placeholders})", val).decode('utf8') for val in values]
            
            # Base sql. We will add to it if we are also doing updates
            sql = f"INSERT INTO {table} ({','.join(list(create_fields))}) VALUES {','.join(mogvals)}"

            # If we are doing update, then build build list of x = excluded.x columns for SET part of sql and add to SQL
            if unique_fields is not None:
                on_duplicates = []
                for field in update_fields:
                    on_duplicates.append(field + "=excluded." + field)
                sql += f" ON CONFLICT ({','.join(list(unique_fields))}) DO UPDATE SET {','.join(on_duplicates)}"

            # Update
            log.debug(f"Saving {len(batch.index)} rows to {table}. Mode: {'INSERT' if not update_fields else 'UPSERT'}.")
            cursor.execute(sql)
            cursor.close()
        else:
            log.debug(f"No data to save.")

## The task code to create a batch and populate the summary tables with aggregated data

### Part 1 - Create the batch, get the data and define the aggregation periods

In [39]:
# Create the batch
batch = models.SummaryBatch(time=timezone.now())
batch.save()

In [40]:
# Get all price data
sql = """SELECT	dscp.id AS datasource_candleperiod_id,
                dss.id AS datasource_symbol_id,
                dss.symbol_id AS symbol_id,
                dscp.period AS period,
                cdl.time AS time
        FROM pricedata_datasourcesymbol dss
                INNER JOIN pricedata_candle cdl
                    ON cdl.datasource_symbol_id = dss.id
                INNER JOIN pricedata_datasourcecandleperiod dscp
                    ON dss.datasource_id = dscp.datasource_id  AND 
                        cdl.period = dscp.period
        WHERE dss.retrieve_price_data = true"""

price_data = pd.read_sql_query(sql=sql, con=connection)

### Part 1b - Define the aggregation periods

In [41]:
# Our summary views will contain min, max and mean aggregatesfor each aggregation period
aggs = {'minutes': 'T', 'hours': 'H', 'days': 'D', 'weeks': 'W', 'months': 'M'}

### Part 2 - Create the 2 summaries, by datasource and cross datasource

In [42]:
# We will create 2 summary views, one by datasource and one across datasources
summary_by_ds_groupby =  ['datasource_symbol_id', 'datasource_candleperiod_id']
symmary_across_ds_groupby =  ['symbol_id', 'period']
group_bys = [summary_by_ds_groupby, symmary_across_ds_groupby]
tables = [models.SummaryMetric.objects.model._meta.db_table, models.SummaryMetricAllDatasources.objects.model._meta.db_table]

for i in range(0,2):
    group_by = group_bys[i]
    table = tables[i]
    
    grouped = price_data.groupby(group_by).agg(first_candle_time=('time', 'min'), last_candle_time=('time', 'max'),
                                               num_candles=('time', 'count'))

    for key in aggs:
        # Get counts for aggregation period, then group by datasource symbol and datasource candleperiod to get min,
        # max and avg counts for each aggregation period
        agg_period_ungrouped = \
            price_data.groupby(group_by + [pd.Grouper(key='time', freq=aggs[key]),]).agg(count=('time', 'count'))

        agg_period_grouped = agg_period_ungrouped.groupby(group_by).agg(
            min=('count', 'min'), max=('count', 'max'), avg=('count', 'median'))

        # Rename columns to include aggregation period key, then merge into original dataframe
        agg_period_grouped = agg_period_grouped.rename(
            columns={'min': f'{key}_min', 'max': f'{key}_max', 'avg': f'{key}_avg'})
        grouped = grouped.join(agg_period_grouped, on=group_by)

    # Reset the grouped index so we end up with a flat dataframe    
    grouped = grouped.reset_index()

    # Add the summary batch id
    grouped['summary_batch_id'] = batch.id
                               
    # Insert into db
    bulk_insert_or_update(data=grouped, table=table)

22:13:11 - __main__ - INFO - Bulk update to pricedata_summarymetric. Rowcount: 228. Update split into 1 batches of 10000 updates.
22:13:11 - __main__ - INFO - Bulk update to pricedata_summarymetric. Batch 1 of 1.
22:15:58 - __main__ - INFO - Bulk update to pricedata_summarymetricalldatasources. Rowcount: 228. Update split into 1 batches of 10000 updates.
22:15:58 - __main__ - INFO - Bulk update to pricedata_summarymetricalldatasources. Batch 1 of 1.


### Part 3 - Create the aggregation / time / symbol view for the dashboard charts

In [51]:
# We will also aggregate the times across each aggregation period and symbol for all datasources and periods. We will only aggregate enough data for each aggregation period for 100 plots
grouped = None
group_by = ['datasource_symbol_id', 'datasource_candleperiod_id']
timedeltas_for_aggs = {'minutes': timedelta(minutes=100), 'hours': timedelta(hours=100), 'days': timedelta(days=100), 'weeks': timedelta(weeks=100), 'months': timedelta(weeks=420)}
for key in aggs:
    # To date is the last date available. From date, depending on aggregation period.
    to_date = price_data['time'].max()
    from_date = to_date - timedeltas_for_aggs[key]
    
    # Filter the data
    filtered = price_data[(price_data['time'] >= from_date) & (price_data['time'] <= to_date)]
    
    # Create grouped for agg period
    agg_grouped = filtered.groupby(group_by + [pd.Grouper(key='time', freq=aggs[key]),]).size().reset_index(name='num_candles')
    
    # Add aggregation period and batch
    agg_grouped['aggregation_period'] = key
    agg_grouped['summary_batch_id'] = batch.id
    
    # Add agg group to grouped
    grouped = agg_grouped if grouped is None else grouped.append(agg_grouped)
  
# Insert into db
bulk_insert_or_update(data=grouped, table=models.SummaryAggregation.objects.model._meta.db_table, batch_size=10000)

22:28:55 - __main__ - INFO - Bulk update to pricedata_summaryaggregation. Rowcount: 29604. Update split into 3 batches of maximum 10000 updates.
22:28:55 - __main__ - INFO - Bulk update to pricedata_summaryaggregation. Batch 1 of 3.
22:28:57 - __main__ - INFO - Bulk update to pricedata_summaryaggregation. Batch 2 of 3.
22:28:58 - __main__ - INFO - Bulk update to pricedata_summaryaggregation. Batch 3 of 3.


## Run queries to display output

In [52]:
# Summary batch
data = pd.read_sql_query(sql="SELECT * FROM pricedata_summarybatch", con=connection)
data

id                             time
0  23 2021-07-29 21:05:11.024469+00:00

In [53]:
# Summary metric
data = pd.read_sql_query(sql="SELECT * FROM pricedata_summarymetric", con=connection)
data

id         first_candle_time          last_candle_time  num_candles  \
0    1825 2021-07-16 00:00:00+00:00 2021-07-29 21:05:03+00:00       265768   
1    1826 2021-07-23 00:06:00+00:00 2021-07-29 21:05:00+00:00         6997   
2    1827 2021-07-23 00:00:00+00:00 2021-07-29 21:00:00+00:00          118   
3    1828 2021-07-16 00:00:00+00:00 2021-07-27 00:00:00+00:00            8   
4    1829 2021-07-16 00:00:00+00:00 2021-07-29 21:05:05+00:00       310220   
..    ...                       ...                       ...          ...   
223  2048 2021-07-23 00:00:00+00:00 2021-07-27 00:00:00+00:00            3   
224  2049 2021-07-16 00:00:00+00:00 2021-07-29 21:05:01+00:00       244494   
225  2050 2021-07-16 00:00:00+00:00 2021-07-29 21:05:00+00:00        14159   
226  2051 2021-07-16 00:00:00+00:00 2021-07-29 21:00:00+00:00          238   
227  2052 2021-07-23 00:00:00+00:00 2021-07-27 00:00:00+00:00            3   

     minutes_min  minutes_max  minutes_avg  hours_min  hours_max  hours_avg  \
0              1           60           17         78       3169        985   
1              1            1            1          6         60         60   
2              1            1            1          1          1          1   
3              1            1            1          1          1          1   
4              1           60           21        114       3125       1241   
..           ...          ...          ...        ...        ...        ...   
223            1            1            1          1          1          1   
224            1           60           16         57       2705       1042   
225            1            1            1          6         60         60   
226            1            1            1          1          1          1   
227            1            1            1          1          1          1   

     ...  days_avg  weeks_min  weeks_max  weeks_avg  months_min  months_max  \
0    ...     25872      24957     135516     105295      265768      265768   
1    ...      1434       1432       5565       3499        6997        6997   
2    ...        24         24         94         59         118         118   
3    ...         1          1          5          2           8           8   
4    ...     30246      30587     154667     124966      310220      310220   
..   ...       ...        ...        ...        ...         ...         ...   
223  ...         1          1          2          2           3           3   
224  ...     24089      22331     122435      99728      244494      244494   
225  ...      1433       1431       7167       5561       14159       14159   
226  ...        24         24        120         94         238         238   
227  ...         1          1          2          2           3           3   

     months_avg  datasource_candleperiod_id  summary_batch_id  \
0        265768                           1                23   
1          6997                           2                23   
2           118                           3                23   
3             8                           4                23   
4        310220                           1                23   
..          ...                         ...               ...   
223           3                           4                23   
224      244494                           1                23   
225       14159                           2                23   
226         238                           3                23   
227           3                           4                23   

     datasource_symbol_id  
0                       1  
1                       1  
2                       1  
3                       1  
4                       2  
..                    ...  
223                    58  
224                    60  
225                    60  
226                    60  
227                    60  

[228 rows x 22 columns]

In [54]:
# Summary metric all datasources
data = pd.read_sql_query(sql="SELECT * FROM pricedata_summarymetricalldatasources", con=connection)
data

id period  minutes_min  minutes_max  minutes_avg  hours_min  hours_max  \
0    1369     1D            1            1            1          1          1   
1    1370     1H            1            1            1          1          1   
2    1371     1M            1            1            1          6         60   
3    1372     1S            1           60           17         78       3169   
4    1373     1D            1            1            1          1          1   
..    ...    ...          ...          ...          ...        ...        ...   
223  1592     1S           19           60           42       1914       3225   
224  1593     1D            1            1            1          1          1   
225  1594     1H            1            1            1          1          1   
226  1595     1M            1            1            1          6         60   
227  1596     1S            1           60           16         57       2705   

     hours_avg  days_min  days_max  ...  weeks_max  weeks_avg  months_min  \
0            1         1         1  ...          5          2           8   
1            1        22        24  ...         94         59         118   
2           60      1261      1435  ...       5565       3499        6997   
3          985     20941     32372  ...     135516     105295      265768   
4            1         1         1  ...          2          2           3   
..         ...       ...       ...  ...        ...        ...         ...   
223       2527     18222     22197  ...     104368      76555      200305   
224          1         1         1  ...          2          2           3   
225          1        22        24  ...        120         94         238   
226         60      1261      1435  ...       7167       5561       14159   
227       1042     21318     28058  ...     122435      99728      244494   

     months_max  months_avg  summary_batch_id  symbol_id  \
0             8           8                23          1   
1           118         118                23          1   
2          6997        6997                23          1   
3        265768      265768                23          1   
4             3           3                23          2   
..          ...         ...               ...        ...   
223      200305      200305                23         58   
224           3           3                23         60   
225         238         238                23         60   
226       14159       14159                23         60   
227      244494      244494                23         60   

            first_candle_time          last_candle_time num_candles  
0   2021-07-16 00:00:00+00:00 2021-07-27 00:00:00+00:00           8  
1   2021-07-23 00:00:00+00:00 2021-07-29 21:00:00+00:00         118  
2   2021-07-23 00:06:00+00:00 2021-07-29 21:05:00+00:00        6997  
3   2021-07-16 00:00:00+00:00 2021-07-29 21:05:03+00:00      265768  
4   2021-07-23 00:00:00+00:00 2021-07-27 00:00:00+00:00           3  
..                        ...                       ...         ...  
223 2021-07-16 10:00:00+00:00 2021-07-29 17:59:59+00:00      200305  
224 2021-07-23 00:00:00+00:00 2021-07-27 00:00:00+00:00           3  
225 2021-07-16 00:00:00+00:00 2021-07-29 21:00:00+00:00         238  
226 2021-07-16 00:00:00+00:00 2021-07-29 21:05:00+00:00       14159  
227 2021-07-16 00:00:00+00:00 2021-07-29 21:05:01+00:00      244494  

[228 rows x 22 columns]

In [55]:
# Summary aggregations
data = pd.read_sql_query(sql="SELECT * FROM pricedata_summaryaggregation", con=connection)
data

id                      time aggregation_period  num_candles  \
0      29508320 2021-07-29 19:25:00+00:00            minutes           11   
1      29508321 2021-07-29 19:26:00+00:00            minutes           11   
2      29508322 2021-07-29 19:27:00+00:00            minutes           25   
3      29508323 2021-07-29 19:28:00+00:00            minutes           22   
4      29508324 2021-07-29 19:29:00+00:00            minutes           28   
...         ...                       ...                ...          ...   
29599  29537919 2021-07-31 00:00:00+00:00             months            3   
29600  29537920 2021-07-31 00:00:00+00:00             months       244494   
29601  29537921 2021-07-31 00:00:00+00:00             months        14159   
29602  29537922 2021-07-31 00:00:00+00:00             months          238   
29603  29537923 2021-07-31 00:00:00+00:00             months            3   

       datasource_candleperiod_id  datasource_symbol_id  summary_batch_id  
0                               1                     1                23  
1                               1                     1                23  
2                               1                     1                23  
3                               1                     1                23  
4                               1                     1                23  
...                           ...                   ...               ...  
29599                           4                    58                23  
29600                           1                    60                23  
29601                           2                    60                23  
29602                           3                    60                23  
29603                           4                    60                23  

[29604 rows x 7 columns]